rustls/src/client/tls12.rs

```diff
@@ -1072,6 +1072,14 @@ impl State<ClientConnectionData> for ExpectTraffic {
             MessagePayload::ApplicationData(payload) => cx
                 .common
                 .take_received_plaintext(payload),
+            MessagePayload::Handshake(HandshakeMessagePayload {
+                typ: HandshakeType::HelloRequest,
+                payload: _,
+            }) => {
+                // XXX(https://github.com/rustls/rustls/issues/952): DoS potential.
+                cx.common
+                    .send_no_renegotiation_warning_alert();
+            }
             _ => {
                 return Err(inappropriate_message(&m, &[ContentType::ApplicationData]));
```

Comment:
This needs to `Err(inappropriate_handshake_message(payload, &[ContentType::Application, ContentType::Handshake], &[HandshakeType::HelloRequest])` (if PR #955 and/or PR #957 are merged), or the more verbose equivalent, if they aren't.